--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -64,6 +64,8 @@
 #include <linux/kcov.h>
 #include <linux/if.h>
 
+#include <net/ra_nat.h>
+
 #include <net/protocol.h>
 #include <net/dst.h>
 #include <net/sock.h>
@@ -2127,6 +2129,9 @@ int pskb_expand_head(struct sk_buff *skb
 	       skb_shinfo(skb),
 	       offsetof(struct skb_shared_info, frags[skb_shinfo(skb)->nr_frags]));
 
+	/*headroom copy*/
+	memcpy(data, skb->head, FOE_INFO_LEN);
+
 	/*
 	 * if shinfo is shared we must drop the old head gracefully, but if it
 	 * is not we can just drop the old head and let the existing refcount
